### PR TITLE
Support for Revit 2022 added

### LINKED ISSUE #1021: Add support for Revit 2022 when applicable
#### Description:

As in the title. Besides adding the necessary solution configs etc., it would be good to refactor all occurences of `#if (REVIT2020 || REVIT2021)` to `#if (REVIT2018 || REVIT2019)` to support the upcoming, higher versions of Revit.

### PR DESCRIPTION
**Pull request by pawelbaran**
### Issues addressed by this PR

Closes #1021

### Test files

Installer available [on SharePoint](https://burohappold.sharepoint.com/sites/BHoM/02_Current/Forms/AllItems.aspx?RootFolder=%2Fsites%2FBHoM%2F02%5FCurrent%2F12%5FScripts%2F02%5FPull%20Request%2FBHoM%2FRevit%5FToolkit%2F%231072%2DRevit2022Support&FolderCTID=0x0120008122C8891F89054B8ACED0196C70DFC4).

### Changelog

- support for Revit 2022 added
- `DisplayUnitType` enum replaced with properties of the `UnitTypeId` static class
- `UnitType` enum replaced with properties of the `SpecTypeId` static class
- `UnitTypeId` and `SpecTypeId` classes implemented for Revit versions older than 2021 to support the changes above (the classes mimic the Revit API ones in Revit 2021 and above)
- `Parameter.DisplayUnitType`  property calls replaced with `GetUnitTypeId` method calls
- `Definition.UnitType`  property calls replaced with `GetSpecTypeId` method calls
- `GetUnitTypeId` and `GetSpecTypeId` methods implemented for Revit versions older than 2021 to support the changes above (the methods mimic the Revit API ones in Revit 2021 and above)

### Additional comments

Do not be afraid of the file change count - most of it is just renaming 🙈 

The changes embrace in this PR are listed in detail in section 1.3 of [this link](https://thebuildingcoder.typepad.com/blog/2020/04/whats-new-in-the-revit-2021-api.html).

### CODE REVIEW

**Comment by pawelbaran** [comment]
@BHoMBot check compliance

**Comment by pawelbaran** [comment]
@BHoMBot check null-handling

**Comment by pawelbaran** [comment]
@BHoMBot check versioning

**Comment by pawelbaran** [comment]
@BHoMBot check core
@BHoMBot check versioning

**Comment by pawelbaran** [comment]
@BHoMBot check compliance

**Comment by pawelbaran** [comment]
@BHoMBot check documentation-compliance

**Comment by pawelbaran** [comment]
@BHoMBot check compliance

**Comment by pawelbaran** [comment]
@BHoMBot check null-handling
@BHoMBot check installer
@BHoMBot check core
@BHoMBot check versioning
@BHoMBot check serialisation

**Comment by pawelbaran** [comment]
I think this is ready for functional review now 👍

**Comment by pawelbaran** [comment]
@BHoMBot check installer
@BHoMBot check core

**Comment by pawelbaran** [comment]
@BHoMBot check installer

**Comment by pawelbaran** [comment]
@BHoMBot check required

File: Revit_Core_Engine/Query/StringValue.cs [hunk 1]
```diff
@@ -31,14 +32,17 @@ public static partial class Query
...
+        [Output("value", "String value extracted from the input Revit Parameter.")]
         public static string StringValue(this Parameter parameter)
         {
             if (parameter != null && parameter.HasValue)
             {
                 switch (parameter.StorageType)
                 {
                     case (StorageType.Double):
-                        return UnitUtils.ConvertFromInternalUnits(parameter.AsDouble(), parameter.DisplayUnitType).ToString();
+                        return UnitUtils.ConvertFromInternalUnits(parameter.AsDouble(), parameter.GetUnitTypeId()).ToString();
```
Comment (alelom) [inline Revit_Core_Engine/Query/StringValue.cs:45]: I think I'm missing something. How does this compile when you set the Solution on any version > Revit2020, if it relies on `GetUnitTypeId()` that becomes unavailable?

Comment (pawelbaran) [reply Revit_Core_Engine/Query/StringValue.cs:45]: The method does not become unavailable - actually, it had been added to the API starting in 2021, so the one implemented in this PR is mimicking the new mechanism. I have just added [this link](https://thebuildingcoder.typepad.com/blog/2020/04/whats-new-in-the-revit-2021-api.html) to the PR description, please feel free to have a look at section 1.3 for the list of changes that I am trying to align to.

**Comment by pawelbaran** [comment]
@BHoMBot check installer

File: Revit_Core_Engine/Query/StringValue.cs (hunk 1 above)
Comment (alelom) [reply Revit_Core_Engine/Query/StringValue.cs:45]: For the record, as from our chat, I thought we should be more explicit on that by changing the description on the GetUnitTypeId() method.

File: Revit_Core_Engine/Query/GetUnitTypeId.cs [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+    public static partial class Query
+    {
+#if (REVIT2018 || REVIT2019 || REVIT2020)
+
+        /***************************************************/
+        /****              Public methods               ****/
+        /***************************************************/
+
+        [Description("A special-purpose method mimicking the Revit API method introduced in 2021 version, which extracts the equivalent of DisplayUnitType from a Revit Parameter.\n" +
+                     "It has been implemented to minimise the fallout caused by the API change on the existing code base of Revit_Toolkit.")]
```
Comment (alelom) [inline Revit_Core_Engine/Query/GetUnitTypeId.cs:38]: As mentioned [here](https://github.com/BHoM/Revit_Toolkit/pull/1072#discussion_r689428712), I'd be more specific in the description.


```suggestion
        [Description("This method is defined by BHoM only for Revit Versions < 2021. Revit versions from 2021 onwards define an equivalent method with the same name as part of their API. This BHoM implementation eliminates breaking changes between different Revit API versions. It returns the equivalent of `DisplayUnitType` from a Revit Parameter."")]
```

**Review by alelom — CHANGES_REQUESTED** [review]
Seems like the best solution we can push for minimizing the impact of Revit 2021 API's breaking changes.

Only a minor suggestion from my side.

**Comment by pawelbaran** [comment]
@BHoMBot check compliance
@BHoMBot check null-handling
@BHoMBot check core
@BHoMBot check versioning
@BHoMBot check serialisation

**Review by alelom — DISMISSED** [review]
Good job, seems to me like the best solution we can possibly implement in order to remedy the breaking changes introduced by the latest Revit API. Approving from a code perspective only.

**Comment by pawelbaran** [comment]
@BHoMBot check required

**Comment by pawelbaran** [comment]
@BHoMBot check required

**Review by michal-pekacki — APPROVED** [review]
Checked in Revit 2018-2021 with success 👍

**Comment by pawelbaran** [comment]
@BHoMBot check copyright-compliance
@BHoMBot check dataset-compliance

**Comment by pawelbaran** [comment]
@BHoMBot check ready-to-merge

**Pull request merged**